src/models/event.js

```diff
@@ -327,17 +347,50 @@ utils.extend(module.exports.MatrixEvent.prototype, {
             );
         }
 
-        try {
-            crypto.decryptEvent(this);
-        } catch (e) {
-            console.warn(
-                `Error decrypting event (id=${this.getId()}): ${e}`,
+        if (this._decrypting) {
+            console.log(
+                `Event ${this.getId()} already being decrypted; queueing a retry`,
             );
+            this._retryDecryption = true;
+            return;
+        }
+
+        this._decrypting = true;
+
+        this._doDecryption(crypto).finally(() => {
+            this._decrypting = false;
+            this._retryDecryption = false;
```

Comment:
Why do we set this to false without inspecting for its truthiness and retrying? I see we do a check in `_doDecryption` but it's only a half retry as we don't retry even if asked to if we fail with a non-DecryptionError. That feels strange and poorly justified.